src/provider.rs

```diff
@@ -117,13 +123,117 @@ impl KBotProvider {
 
 #[async_trait]
 impl ModelProvider for KBotProvider {
-    async fn get_joint_angles(
+    async fn get_inputs(
+        &self,
+        input_types: &[InputType],
+        meta: &ModelMetadata,
+    ) -> Result<HashMap<InputType, Array<f32, IxDyn>>, ModelError> {
+        use InputType::*;
+
+        // Read values from hardware once
+        let actuator_ids = self.get_actuator_ids(&meta.joint_names)?;
+        let (act_state, imu_values) =
+            tokio::try_join!(self.get_actuator_state(&actuator_ids), async {
+                self.imu.get_values().await.map_err(|e| {
+                    ModelError::Provider(format!("Failed to get IMU values: {}", e.to_string()))
+                })
+            })?;
```

Comment:
i think something like this is better

```suggestion
        let (act_state, imu_values) =
            tokio::try_join!(async {
                let actuator_ids = self.get_actuator_ids(&meta.joint_names)?;
                self.get_actuator_state(&actuator_ids).await
            },
            async {
                self.imu.get_values().await.map_err(|e| {
                    ModelError::Provider(format!("Failed to get IMU values: {}", e.to_string()))
                })
            })?;
```